### PR TITLE
Types: Factor out 'shape' producers in preparation for schema

### PR DESCRIPTION
**Pull request by DLehenbauer**

### CODE REVIEW

File: .devcontainer/.bashrc [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# ~/.bashrc: executed by bash(1) for non-login shells.
```
Comment (DLehenbauer) [inline .devcontainer/.bashrc:1]: Sorry, some unrelated devcontainer changes leaked into my branch.  Ignore the first few files.

File: packages/core/micro/src/map/producer.ts [hunk 2]
```diff
@@ -6,28 +11,28 @@ export abstract class Producer<TMap> implements IProducer<TMap>, IReader<TMap> {
...
-    public get producer() { return this; }
+    public get producer():IProducer<TMap> { return this; }
 
     //#endregion IReader
 
-    protected invalidateValue<K extends keyof TMap>(property: K): void {
+    protected invalidateValue<K extends keyof TMap>(key: K): void {
         forEachConsumer(this.consumers, (consumer) => {
-            consumer.valueChanged(property, this);
+            consumer.keyChanged(key, this);
```
Comment (DLehenbauer) [inline packages/core/micro/src/map/producer.ts:35]: The 'key' terminology works well for records & sets, both when observing the "shape" and the "values":

IConsumer {
    // Value associated with the key has changed
    keyChanged(key: string, producer: IShapeProducer
}

IShapeConsumer {
    // The key has been added or removed
    keyChanged(key: string, producer: IShapeProducer
}

File: packages/core/micro/test/matrix.spec.ts [hunk 3]
```diff
@@ -95,8 +102,8 @@ export class TestMatrix<T = any, TRow = never, TCol = never> implements IMatrixC
 
     public constructor (
         producer: IMatrixProducer<T>,
-        private readonly rowWriter: IVectorWriter<TRow>,
-        private readonly colWriter: IVectorWriter<TCol>,
+        private readonly rowWriter: IVectorWriter<TRow> & IVectorShapeWriter,
+        private readonly colWriter: IVectorWriter<TCol> & IVectorShapeWriter,
```
Comment (DLehenbauer) [inline packages/core/micro/test/matrix.spec.ts:106]: Writers can overwrite existing values.
ShapeWriters can move, remove, and insert empty items.

In general, if you need to know <T> for the operation, it's a write.
If you can do the operation w/o knowing T, it's a shape change.

File: packages/types/es/README.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
+# @tiny-calc/types
```
Comment (DLehenbauer) [inline packages/types/es/README.md:1]: Heh.  Copy & pasted wrong readme.

File: packages/types/es/src/produce.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
+/*!
```
Comment (DLehenbauer) [inline packages/types/es/src/produce.ts:1]: This is 75% move and 25% update to support IShape*

File: packages/types/types/src/record.ts [hunk 6]
```diff
@@ -3,45 +3,102 @@
...
+    readonly producer?: IShapeProducer<TKey>;
+}
+
+export interface IShapeWriter<TKey> {
+    delete(key: TKey): void;
+    clear(): void;
+}
+
 /** An observable key/value collection, such as a Record or Map. */
 export interface IProducer<T> {
```
Comment (DLehenbauer) [inline packages/types/types/src/record.ts:50]: Our base 'IProducer' is the only variant that isn't implicitly an IShapeProducer.

IMapProducer (below) is a convenient way to implement both IProducer and IShapeProducer

File: packages/types/types/src/matrix.ts [hunk 7]
```diff
@@ -4,49 +4,70 @@
...
+    spliceRows(rowStart: number, deleteCount: number, insertCount: number): void;
+    spliceCols(colStart: number, deleteCount: number, insertCount: number): void;
 }
 
-/** A consumer of change notifications for a matrix. */
-export interface IMatrixConsumer<T> {
-    /** Notification that rows have been inserted, removed, and/or replaced in the given matrix. */
-    rowsChanged(rowStart: number, removedCount: number, insertedCount: number, producer: IMatrixProducer<T>): void;
+/** An observable 2D numerically indexed collection. */
+export interface IMatrixProducer<T> extends IMatrixShapeProducer {
```
Comment (DLehenbauer) [inline packages/types/types/src/matrix.ts:52]: Generally, an I*Producer is implicitly an I*ShapeProducer, an I*Reader is an I*ShapeReader, and so on...

The exceptions are:
* I*Writers are not I*ShapeWriters.   (e.g., the capability to set cells in a matrix is separate from the capability to resize a matrix.)
* IProducer is not an IShapeProducer because not all IProducers support reflection.  (At least, currently.)

File: .devcontainer/.bashrc (hunk 1 above)
Comment (DLehenbauer) [reply .devcontainer/.bashrc:1]: ...actually...  all the good stuff (@tiny-calc/types changes) are at the bottom.  Maybe start there?

The main idea is to split I*Producer into two parts, one that describes the shape and one that describes the values...  just like the ITree* interfaces.

File: packages/types/es/src/produce.ts [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    // #region IShapeReader
+    keys:  { value: function() { return Object.keys(this); }},
+    has:   { value: function (key: any) { return Object.keys(this).includes(key); }},
+    size:  { get: function() { return Object.keys(this).length; }},
+    // #endregion IShapeReader
+}
+
+const vectorProps = {
+    // #region IProducer, IReader, IShapeReader
+    ...props,
```
Comment (jack-williams) [inline packages/types/es/src/produce.ts:32]: Is this approach faster than a class? Looks clean though.

File: packages/types/types/src/matrix.ts (hunk 7 above)
Comment (jack-williams) [reply packages/types/types/src/matrix.ts:52]: The co and contra variance between reading and writing feels pretty natural.

Value reading implies shape reading; I need to know the shape to read a value.
Shape writing implies value writing; to change the shape I need to create or destroy exiting values.

but,

I can read a shape without reading the values, and I can write values without writing the shape.

File: packages/types/types/src/record.ts [hunk 9]
```diff
@@ -3,45 +3,102 @@
...
      * or undefined if the producer is immutable.
      */
-    readonly producer?: IProducer<T>;
+    readonly producer?: IProducer<T> | (IProducer<T> & IShapeProducer<keyof T>);
 }
 
 /** Capability to set values of a key/value collection. */
 export interface IWriter<T> {
-    set<K extends keyof T>(property: K, value: T[K]): void;
+    set<K extends keyof T>(key: K, value: T[K]): void;
```
Comment (jack-williams) [inline packages/types/types/src/record.ts:86]: What is the contract around setting keys not current in `T`. Will that create a new mapping, or fail?

File: packages/types/types/src/record.ts [hunk 10]
```diff
@@ -3,45 +3,102 @@
...
+    readonly size: number;
+
+    /**
+     * A reference to the underlying producer that provides the shape for this reader,
+     * or undefined if the producer is immutable.
+     */
+    readonly producer?: IShapeProducer<TKey>;
+}
+
+export interface IShapeWriter<TKey> {
```
Comment (jack-williams) [inline packages/types/types/src/record.ts:44]: Do new keys ever get introduced?

File: packages/types/types/src/record.ts [hunk 11]
```diff
@@ -3,45 +3,102 @@
...
     close(consumer: IConsumer<T>): void;
 }
 
+/** A consumer of change notifications for a key/value collection, such as a Record or Map. */
+export interface IConsumer<T> {
+    keyChanged<K extends keyof T>(key: K, producer: IProducer<T>): void;
+}
+
 /** Capability to read values of a key/value collection. */
 export interface IReader<T> {
```
Comment (jack-williams) [inline packages/types/types/src/record.ts:73]: I think we've been implicitly assuming that `T` is object-like, rather than a primitive. I'm not sure it is worth putting this in the constraint, it depends on the dev experience. I think it should be left as-is, but I'd be interested to see if we find any practical reasons for/against.

**Review by jack-williams — APPROVED** [review]
Looks cool!

File: packages/types/types/src/matrix.ts (hunk 7 above)
Comment (DLehenbauer) [reply packages/types/types/src/matrix.ts:52]: There's one subtly when shapes are shared.  For example, I create a resizable matrix like this:
```ts
const values: IMatrixWriter & IMatrixShapeWriter = new ResizableMatrix();
```
And someone else creates a matrix that shares the same shape:
```ts
const annotations: IMatrixWriter = new Matrix(/* shapeProducer: */ values);
```
In this situation 'annotations' is an IMatrixWriter, but not an IShapeMatrixWriter.  This means that value producers need to handle key insertion w/o an associated value by producing an empty/default value like 'undefined'.

This does create a "glitch" between key insertion and setting the values:
```ts
values.insertRows(0, 10);    // inserts 10 empty rows
values.setValues(0, 0, /* colCount: */ 3, [ ... ]);  // add associated values 
annotations.setValues(0, 0, /* colCount: */ 3, [ ... ]);  // add associated annotations 
```
IMO, the way to solve this is some form of implicit transaction, such as lazy recalc or deferring/scheduling recalc for the next microtask.

Value reading does imply shape reading, except for "records", which are maps with an immutable shape.  My thinking was that since the shape is immutable, we shouldn't require the implementer to provide 'keys()'/'has()'...  but it's a bit of an experiment.  I can easily see us backtracking on that one.

File: packages/types/es/src/produce.ts (hunk 8 above)
Comment (DLehenbauer) [reply packages/types/es/src/produce.ts:32]: Yeah, I feel dirty about it...  but the perf is excellent:
* For map-like objects, reader.get("bar") is ~ as fast as reader["bar"]
* For arrays, reader.getItem(43) is ~ as fast as reader[43]

Probably the most common usage of 'produce' will be combined with JSON parse, in which case we can wrap it up like this so that no one can get their hands on the mutable intermediate objects:
```ts
const producer: IProducer = Producer.parse("{ a: 1, b: 2 }");
```
...or we can have "debug" builds where produce(..) implicitly freezes the source objects to help make this pattern safer.

File: packages/types/types/src/record.ts (hunk 9 above)
Comment (DLehenbauer) [reply packages/types/types/src/record.ts:86]: It's disallowed by the contract, with the caveat that you can define T to allow arbitrary keys:
```ts
const writer: IWriter<Record<string, any>> = new Map<string, any>();
writer.set("a", 1);    // allowed because "a" satisfies keyof Record<string, any>
```
(However, I don't expect producers to guard against setting adding keys this way at runtime.)

File: packages/types/types/src/record.ts (hunk 10 above)
Comment (DLehenbauer) [reply packages/types/types/src/record.ts:44]: Strictly, no...  but you can leave the set of keys open using `IWriter<Record<string, any>>`...

It occurs to me now that I should probably stop saying "records are things with an immutable shape" since I use 'Record<TKey, TValue>' to allow the shape to mutate.  Have to come up with another term...

File: packages/types/types/src/record.ts (hunk 11 above)
Comment (DLehenbauer) [reply packages/types/types/src/record.ts:73]: Oh yeah...  you're right.

One consideration: right now I'm assuming 'Object.keys(instanceOfT)' and 'keyof T' contain the same  set of keys, but that's only true if 'T' is a plain object.  For something like a string, 'keyof T' will contain a lot of extra stuff like 'toString()', 'indexOf()', 'toLowerCase()', etc.
